udata/core/spam/models.py

```diff
@@ -0,0 +1,161 @@
+from flask import current_app
+from mongoengine import signals
+from langdetect import detect
+
+from udata.models import db
+from .signals import on_new_potential_spam
+
+NOT_CHECKED = 'not_checked'
+POTENTIEL_SPAM = 'potentiel_spam'
+NO_SPAM = 'no_spam'
+
+class SpamInfo(db.EmbeddedDocument):
+    status = db.StringField(choices=[NOT_CHECKED, POTENTIEL_SPAM, NO_SPAM], default=NOT_CHECKED)
+    callbacks = db.DictField(default={})
+
+class SpamMixin(object):
+    spam = db.EmbeddedDocumentField(SpamInfo)
+
+    @staticmethod
+    def spam_words():
+        return current_app.config.get('SPAM_WORDS', [])
+    
+    @staticmethod
+    def allowed_langs():
+        return current_app.config.get('SPAM_ALLOWED_LANGS', [])
+
+    def detect_spam_listener(sender, document, *args, **kwargs):
```

Comment:
The problem with the `clean` method is that it's using inheritance and with multiple inheritance it may become weird… For example if we have two mixins that want to do something on `clean()`

Reply:
I'm unsure about the method resolution order, but you can look that up, but in any case, they should all be called :)

Reply:
In my test, they are not all called… But maybe I'm wrong somewhere… In the following example, only the first mixin is called with a `super()`

```python
class MixinA:
    def test(self):
        print("Inside Mixin A")

class MixinB:
    def test(self):
        print("Inside Mixin B")

class Main(MixinA, MixinB):
    def test(self):
        super(Main, self).test()
        print("Inside Main")

Main().test()
```

```
(venv) ➜  udata git:(improved_spam_prevention) ✗ python test.py
Inside Mixin A
Inside Main
```